lib/charms/smtp_integrator/v0/smtp.py

```diff
@@ -0,0 +1,277 @@
+        transport_security: The security protocol to use for the outgoing SMTP relay.
+        domain: The domain used by the sent emails from SMTP relay.
+    """
+
+    @property
+    def host(self) -> str:
+        """Fetch the SMTP host from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("host")
+
+    @property
+    def port(self) -> int:
+        """Fetch the SMTP port from the relation."""
+        assert self.relation.app
+        return int(self.relation.data[self.relation.app].get("port"))
+
+    @property
+    def user(self) -> str:
+        """Fetch the SMTP user from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("user")
+
+    @property
+    def password(self) -> str:
+        """Fetch the SMTP password from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("password")
+
+    @property
+    def auth_type(self) -> str:
+        """Fetch the SMTP auth type from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("auth_type")
+
+    @property
+    def transport_security(self) -> str:
+        """Fetch the SMTP transport security protocol from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("transport_security")
+
+    @property
+    def domain(self) -> str:
+        """Fetch the SMTP domain from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("domain")
+
+
+class SmtpRequiresEvents(ops.CharmEvents):
+    """SMTP events.
+
+    This class defines the events that a SMTP requirer can emit.
+
+    Attrs:
+        smtp_data_available: the SmtpDataAvailableEvent.
+    """
+
+    smtp_data_available = ops.EventSource(SmtpDataAvailableEvent)
+
+
+class SmtpRequires(ops.Object):
+    """Requirer side of the SMTP relation.
+
+    Attrs:
+        on: events the provider can emit.
+    """
+
+    on = SmtpRequiresEvents()
+
+    def __init__(self, charm: ops.CharmBase, relation_name: str = DEFAULT_RELATION_NAME) -> None:
+        """Construct.
+
+        Args:
+            charm: the provider charm.
+            relation_name: the relation name.
+        """
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+
+    def _on_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        """Event emitted when the relation has changed.
+
+        Args:
+            event: event triggering this handler.
+        """
+        assert event.relation.app
+        if event.relation.data[event.relation.app]:
+            self.on.smtp_data_available.emit(event.relation, app=event.app, unit=event.unit)
```

Comment:
I think the data should be validated here too

Reply:
You mean the relation content? It's already validated before converting it to strings

Reply:
But it's validated on the Provider side, not the required. And we have no guarantee that the Provider library was used when consuming the data from the Requirer. So as a rule of thumbs, always validate the data from both sides

Reply:
Perhaps we should instantiate a `SmtpRelationData` object here and merge this object with `SmtpDataAvailableEvent`. Otherwise we'd be duplicating the code

Reply:
That's fine, indeed better to avoid duplicating the validation

Reply:
Sorry re-opening, we probably misunderstood. When I said "that's fine" I was approving your approach of instantiating a `SmtpRelationData` object here to trigger validation

Reply:
The approach is a bit different. I'm instantiating it when accessing the event data instead of when emitting it for two reasons:
* I don't want to override the ops `emit`method
* This seems more cohesive. The event will be emitted in more places in the future
Let me know your thoughts on this

Reply:
Changed to instantiate immediately before emitting